File: apps/web/types/budget.ts
/**
 * Budget-related types matching backend DTOs
 *
 * These types correspond to the budget module in services/api/app/schemas/budget.py
 * All types use snake_case to match backend DTOs exactly.
 */

/**
 * Budget meter data showing spending progress for an occasion
 */
export interface BudgetMeterData {
  budget_total: number | null;
  purchased_amount: number;
  planned_amount: number;
  remaining_amount: number | null;
  purchased_percent: number;
  planned_percent: number;
  is_over_budget: boolean;
  has_budget: boolean;
}

/**
 * Budget warning levels and thresholds
 */
export interface BudgetWarning {
  level: 'none' | 'approaching' | 'near_limit' | 'exceeded';
  message: string;
  threshold_percent: number;
  current_percent: number;
}

/**
 * Entity-specific budget (list or person within an occasion)
 */
export interface EntityBudget {
  entity_type: string;
  entity_id: number;
  budget_amount: number;
  spent_amount: number;
  remaining_amount: number;
  is_over_budget: boolean;
}

/**
 * Request to set/update occasion budget
 */
export interface SetBudgetRequest {
  budget_amount: number;
}

/**
 * Request to set/update entity-specific budget
 */
export interface SetEntityBudgetRequest {
  entity_type: string;
  entity_id: number;
  budget_amount: number;
}

/**
 * Person budget calculation showing gifts assigned to and purchased by a person
 */
export interface PersonBudget {
  person_id: number;
  occasion_id: number | null;

  // Recipient role (gifts TO this person)
  gifts_assigned_count: number;
  gifts_assigned_total: number;  // Decimal serialized as float

  // NEW: Of assigned gifts, how many are purchased
  gifts_assigned_purchased_count: number;
  gifts_assigned_purchased_total: number;  // Decimal serialized as float

  // Purchaser role (gifts BY this person - already purchased)
  gifts_purchased_count: number;
  gifts_purchased_total: number; // Decimal serialized as float

  // NEW: Gifts assigned as purchaser but not yet bought
  gifts_to_purchase_count: number;
  gifts_to_purchase_total: number;  // Decimal serialized as float
}

File: services/api/app/schemas/budget.py
"""Budget DTOs for budget tracking and progression meters."""

from decimal import Decimal

from pydantic import BaseModel, Field

from app.schemas.base import BaseSchema


class BudgetMeterDTO(BaseSchema):
    """Budget meter data for frontend display."""

    budget_total: Decimal | None = Field(
        None,
        description="Total budget for the occasion (None if not set)",
    )
    purchased_amount: Decimal = Field(
        default=Decimal("0"),
        description="Total amount spent on purchased items",
    )
    planned_amount: Decimal = Field(
        default=Decimal("0"),
        description="Total amount for reserved/planned items",
    )
    remaining_amount: Decimal | None = Field(
        None,
        description="Remaining budget (None if no budget set)",
    )

    # Calculated percentages (for progress bar display)
    purchased_percent: float = Field(
        default=0.0,
        ge=0.0,
        description="Percentage of budget that's been purchased",
    )
    planned_percent: float = Field(
        default=0.0,
        ge=0.0,
        description="Percentage of budget that's planned/reserved",
    )
    remaining_percent: float = Field(
        default=100.0,
        ge=0.0,
        description="Percentage of budget remaining",
    )

    # Status flags
    is_over_budget: bool = Field(
        default=False,
        description="True if total committed exceeds budget",
    )
    has_budget: bool = Field(
        default=False,
        description="True if a budget is set for this occasion",
    )


class BudgetWarningDTO(BaseModel):
    """Warning information when approaching or exceeding budget."""

    level: str = Field(
        ...,
        description="Warning level: 'none', 'approaching', 'near_limit', 'exceeded'",
    )
    message: str = Field(
        ...,
        description="Human-readable warning message",
    )
    threshold_percent: float = Field(
        ...,
        description="The threshold percentage that triggered this warning",
    )
    current_percent: float = Field(
        ...,
        description="Current usage percentage",
    )


class EntityBudgetDTO(BaseSchema):
    """Budget information for a specific entity (gift or list)."""

    entity_type: str = Field(
        ...,
        description="'gift' or 'list'",
    )
    entity_id: int = Field(
        ...,
        description="ID of the gift or list",
    )
    budget_amount: Decimal | None = Field(
        None,
        description="Allocated budget (None if not set)",
    )
    spent_amount: Decimal = Field(
        default=Decimal("0"),
        description="Amount spent on this entity",
    )
    remaining_amount: Decimal | None = Field(
        None,
        description="Remaining budget (None if no budget set)",
    )
    is_over_budget: bool = Field(
        default=False,
        description="True if spent exceeds allocated budget",
    )


class SetBudgetRequest(BaseModel):
    """Request to set/update a budget."""

    budget_amount: Decimal = Field(
        ...,
        gt=0,
        description="Budget amount (must be positive)",
    )


class SetEntityBudgetRequest(BaseModel):
    """Request to set/update an entity budget."""

    entity_type: str = Field(
        ...,
        description="'gift' or 'list'",
    )
    entity_id: int = Field(
        ...,
        description="ID of the gift or list",
    )
    budget_amount: Decimal = Field(
        ...,
        gt=0,
        description="Budget amount (must be positive)",
    )
